### PR TITLE
Introduce Functions\stubWpUrlFunctions() helper

### PR DESCRIPTION
**Pull request by gmazzap**
It makes use of a new UrlsHelper class.

Tests and documentation added.

See #125

### CODE REVIEW

File: docs/functions-testing-tools/function-stubs.md [hunk 1]
```diff
@@ -141,6 +141,40 @@ Only for functions that both translate and escape \(`esc_html__()`, `esc_html_x(
...
+* `network_home_url()`
+* `network_site_url()`
+* `network_admin_url()`
+* `user_admin_url()`
+* `wp_login_url()`
+
+The function accepts as first argument the domain to use for subbing the URLs, default to `example.org`.
+
+E.g., after `Functions\stubWpUrlFunctions()`, the `home_url()` function returns `https://example.org`,
+but after `Functions\stubWpUrlFunctions('acme.com')`, the `home_url()` function returns `https://acme.org`.
```
Comment (jrfnl) [inline docs/functions-testing-tools/function-stubs.md:172]:
```suggestion
E.g., calling `Functions\stubWpUrlFunctions()`, the `home_url()` function returns `https://example.org`,
but calling `Functions\stubWpUrlFunctions('acme.com')`, the `home_url()` function returns `https://acme.org`.
```

The word "after" could give the impression that some form of caching/remembering of previously passed function call arguments is being done. This small change removes that confusion.

File: inc/api.php [hunk 2]
```diff
@@ -182,6 +183,39 @@ function stubEscapeFunctions()
...
+            'rest_url' => $helper->stubUrlCallback('wp-json'),
+            'get_rest_url' => $helper->stubUrlForSiteCallback('wp-json'),
+            'includes_url' => $helper->stubUrlCallback('wp-includes'),
+            'network_home_url' => $helper->stubUrlCallback(),
+            'network_site_url' => $helper->stubUrlCallback(),
+            'network_admin_url' => $helper->stubUrlCallback('wp-admin/network', 'admin'),
+            'user_admin_url' => $helper->stubUrlCallback('wp-admin/user', 'admin'),
+            'wp_login_url' => static function ($redirect = '', $force_reauth = false) use ($helper) {
+                $callback = $helper->stubUrlCallback();
+                $url = $callback('/wp-login.php', 'login');
```
Comment (jrfnl) [inline inc/api.php:211]: Small simplification as the `$callback` variable isn't used anywhere else within the closure.

```suggestion
                $url = $helper->stubUrlCallback('/wp-login.php', 'login');
```

File: inc/api.php [hunk 3]
```diff
@@ -182,6 +183,39 @@ function stubEscapeFunctions()
...
+                $has_redirect = ($redirect !== '') && is_string($redirect);
+                $has_redirect and $url .= '?redirect_to=' . urlencode($redirect);
+                $force_reauth and $url .= ($has_redirect ? '&reauth=1' : '?reauth=1');
```
Comment (jrfnl) [inline inc/api.php:214]: This code has high cognitive load due to the use of `and`.

Two alternative ways to write the same while lowering the cognitive load:
```suggestion
                $url .= $has_redirect ? '?redirect_to=' . urlencode($redirect) : '';
                $url .= $force_reauth ? ($has_redirect ? '&reauth=1' : '?reauth=1') : '';
```

```suggestion
                if ($has_redirect) {
                    $url .= '?redirect_to=' . urlencode($redirect);
                }
                if ($force_reauth) {
                    $url .= ($has_redirect ? '&reauth=1' : '?reauth=1');
                }
```

File: src/Expectation/UrlsHelper.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
+<?php
```
Comment (jrfnl) [inline src/Expectation/UrlsHelper.php:1]: All `@param` declarations in this file are missing the `type` information, which makes them kind of useless.

File: inc/api.php [hunk 5]
```diff
@@ -182,6 +183,39 @@ function stubEscapeFunctions()
...
+            'get_admin_url' => $helper->stubUrlForSiteCallback('wp-admin', 'admin'),
+            'content_url' => $helper->stubUrlCallback('wp-content', null, false),
+            'rest_url' => $helper->stubUrlCallback('wp-json'),
+            'get_rest_url' => $helper->stubUrlForSiteCallback('wp-json'),
+            'includes_url' => $helper->stubUrlCallback('wp-includes'),
+            'network_home_url' => $helper->stubUrlCallback(),
+            'network_site_url' => $helper->stubUrlCallback(),
+            'network_admin_url' => $helper->stubUrlCallback('wp-admin/network', 'admin'),
+            'user_admin_url' => $helper->stubUrlCallback('wp-admin/user', 'admin'),
+            'wp_login_url' => static function ($redirect = '', $force_reauth = false) use ($helper) {
```
Comment (jrfnl) [inline inc/api.php:209]: It looks like most of these don't take passed parameters to the original function into account, while WP does support passing parameters.

Stubbing these in the framework without taking the passes parameters into account will probably be confusing and lead to support overhead or to people just not using these stubs as they don't do what they expect.

File: src/Expectation/UrlsHelper.php (hunk 4 above)
Comment (gmazzap) [reply src/Expectation/UrlsHelper.php:1]: All `@param` are mixed anyway. The type is checked inside the function via `is_string` and similar. So you can pass whatever. Mixed is implicit so to me that's fine. 
If we would add Psalm or similar, as soon I would declare `@param string` and then do `is_string()` checks inside the function Psalm would rightfully complain for "redundant check". 
I prefer to keep the check at runtime so all the parameters are mixed, and IMO there's no point in typing "mixed" all over the place.
Still, I like to keep the doc block for "visual help". I personally find it hard to digest a wall of code without some "calming" doc block in between functions.

File: src/Expectation/UrlsHelper.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    private $use_https;
+
+    /**
+     * @param $domain
+     * @param $use_https
+     */
+    public function __construct($domain = 'example.org', $use_https = null)
+    {
+        $this->domain = (is_string($domain) && $domain) ? $domain : 'example.org';
```
[the remark this answers is not in the document]
Comment (gmazzap) [reply src/Expectation/UrlsHelper.php]: With just point 1, I would need to define a default value for the parameter anyway, to allow people to construct the helper without passing any parameter. 

So I went with point 2. and the default value of the parameter is now null, and the property is set to either a valid provided value or to the default in the constant.

File: inc/api.php (hunk 2 above)
Comment (gmazzap) [reply inc/api.php:211]: @jrfnl `stubUrlCallback()` return a *callback that returns a string*, not a string.
Following your suggestion `$url` would be a closure, not a string.
The simplification would be:  `$url = $helper->stubUrlCallback()('/wp-login.php', 'login');` but that does not work on PHP 5.6 where this lib is still stuck.

File: inc/api.php (hunk 5 above)
Comment (gmazzap) [reply inc/api.php:209]: @jrfnl I think you need to look at it again :)

`$helper->stubUrlCallback()` and `$helper->stubUrlForSiteCallback()` return _callbacks_ that take respectively 2 (`$path`, `$schema`) and 3 parameters (`$blog_id`, `$path`, `$schema`), that are **the same core functions take**.

There are 2 exceptions: `content_url()` and `wp_login_url()`.

`wp_login_url()` takes still 2 arguments, but different from the others, and that is why it is stubbed differently.

`content_url()` takes 1 parameter: `$path`. That is why `stubUrlCallback()` accepts a 3rd parameter `$use_schema_arg` that defaults to `true`, but it is passed as `false` for `content_url()`.

To be noted:

- the `$blog_id` parameter for functions that accepts it is just ignored in stubbed functions because the base URL is taken from what is passed to the `stubWpUrlFunctions()` API function.
- When `$schema` param is something like `'admin'` or `'login'` it is taken into account only if HTTPs is not forced (passing `true` as `stubWpUrlFunctions()`'s 2nd parameter) and `force_ssl_admin()` is available (likely because stubbed separately).
- `rest_url()` and `get_rest_url()`  use `'rest'` as scheme. That is not relevant for the generation of the URL (in WP just like in the stubbed version), it is used in WP only as the 3rd parameter, `$orig_scheme`, passed to the `'set_url_scheme'` hook, something that I think is pretty safe to ignore for the stubbed version of the function.

**TL;DR**:  all the stubbed functions accept the same parameters and work very similarly to the WP counterparts. For some edge cases, it might be needed to stub `is_ssl()` and/or `force_ssl_admin()` separately.

**Comment by gmazzap** [comment]
At this point, I think the only thing missing before this can be merged is **tests**.

**Comment by gmazzap** [comment]
> I've not completely reviewed this PR yet as I believe this remark needs to be addressed/hashed out first:
> 
> > It looks like most of these don't take passed parameters to the original function into account, while WP does support passing parameters.
> > Stubbing these in the framework without taking the passes parameters into account will probably be confusing and lead to support overhead or to people just not using these stubs as they don't do what they expect.
> 
> As addressing that will require some rework in the PR (if it will be addressed), I figured my time would be better spend reviewing after the discussion has been had about the above.

Commented al your points. Your concerns about parameters, as better explained below, isn't correct because parameters *are* accepted in the same way WP does. Besides my reasoning in other comments, there're tests to prove it :)

File: inc/api.php (hunk 2 above)
Comment (jrfnl) [reply inc/api.php:211]: Darn, of course!
